service/s3/api_op_GetBucketLocation.go

```diff
@@ -133,6 +143,62 @@ func addOperationGetBucketLocationMiddlewares(stack *middleware.Stack, options O
 	return nil
 }
 
+type awsRestxml_deserializeOpGetBucketLocation_custom struct {
+}
+
+func (*awsRestxml_deserializeOpGetBucketLocation_custom) ID() string {
+	return "OperationDeserializer"
+}
+
+func (m *awsRestxml_deserializeOpGetBucketLocation_custom) HandleDeserialize(ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
+	out middleware.DeserializeOutput, metadata middleware.Metadata, err error,
+) {
+	out, metadata, err = next.HandleDeserialize(ctx, in)
+	if err != nil {
+		return out, metadata, err
+	}
+
+	response, ok := out.RawResponse.(*smithyhttp.Response)
+	if !ok {
+		return out, metadata, &smithy.DeserializationError{Err: fmt.Errorf("unknown transport type %T", out.RawResponse)}
+	}
+
+	if response.StatusCode < 200 || response.StatusCode >= 300 {
+		return out, metadata, awsRestxml_deserializeOpErrorGetBucketLocation(response, &metadata)
+	}
+	output := &GetBucketLocationOutput{}
+	out.Result = output
+
+	var buff [1024]byte
+	ringBuffer := smithyio.NewRingBuffer(buff[:])
+	body := io.TeeReader(response.Body, ringBuffer)
+	rootDecoder := xml.NewDecoder(body)
+	decoder := smithyxml.WrapNodeDecoder(rootDecoder, xml.StartElement{})
+	err = awsRestxml_deserializeOpDocumentGetBucketLocationOutput(&output, decoder)
+	if err == io.EOF {
+		err = nil
+	}
+	if err != nil {
+		var snapshot bytes.Buffer
+		io.Copy(&snapshot, ringBuffer)
+		return out, metadata, &smithy.DeserializationError{
+			Err:      fmt.Errorf("failed to decode response body, %w", err),
+			Snapshot: snapshot.Bytes(),
+		}
+	}
+
+	return out, metadata, err
+}
+
+// $L is helper to swap in a custom deserializer
+func swapDeserializerHelper(stack *middleware.Stack) error {
+	_, err := stack.Deserialize.Swap("OperationDeserializer", &awsRestxml_deserializeOpGetBucketLocation_custom{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
is there any reasonable way this can be done in the code generator?  Replace at codegen time vs runtime is something we should look at how the codegen can be updated to support custom (de)serializers.

Don't have to do that here, but we should be looking at ways the codegen can be updated to support that.

Reply:
Since middlewares are added at runtime, replacing them at runtime made sense. Customizing the code generator is a larger change, but not needed.

Reply:
One benefit of code generator taking care of this would be to prevent runtime allocations churn that never would be needed, and dead code that would never be used. Basically cleaner and less overhead for something that will never be used.

Also agree, right now this is probably too big of a change to take today, but we should be thinking about this long term.